### PR TITLE
Adapt the Meson Build System

### LINKED ISSUE #27: Tidy up build process
* [x] Build on Windows and Linux
* [x] Rewrite Makefile (or use alternative build) to be independent of system specifics
* [ ] Add Install.md
* [x] Support latest wxWidgets

### PR DESCRIPTION
**Pull request by mkroening**
In an effort to build MiniPlaner on Linux, I figured that instead of fiddling around with the makefile I'd rather try to port the project to [The Meson Build System](https://mesonbuild.com/). That way the build files should be easier to read and it should work better on different platforms.

- [x] verify functionality on Windows
   - [x] compile resources with icon, resolves https://github.com/yannikschaelte/MiniPlaner/issues/7
   - [x] compile statically
   - [x] soft code resource compiler arguments, see https://github.com/mesonbuild/meson/issues/6897
- [x] resolve an issue with selecting the GTK version, see https://github.com/mesonbuild/meson/issues/6811

What do you think?

### CODE REVIEW

**Comment by mkroening** [comment]
I succeeded in building a statically linked binary on MSYS2 with Meson.

I created [mwkroening/MiniPlaner-MINGW](https://github.com/mwkroening/MiniPlaner-MINGW) to house the package scripts for that environment. See the impl branch for a WIP version of the package. I will merge the package script once the required changes have been merged to this repository.

What do you think, @yannikschaelte?

**Comment by mkroening** [comment]
Rebased on #31, since we need the Windows resources.

**Comment by mkroening** [comment]
This is ready for review! Apart on how we handle the README (see #34), I am quite happy with this.

@yannikschaelte, @janochsner, could you try this out? That way we can further improve documentation on compilation.

**Comment by yannikschaelte** [comment]
Looks awesome, thanks a lot! I ran it on Linux with wxWidgets 3.0.4, it worked just fine. Some questions:

* I don't currently have access to Windows, but can for the moment someone confirm that it works on Windows 8 or 10?
* I have a binary size of 18MB. On Windows, it used to be 7MB. It is probably just the Linux build, or has something in the linking changed s.t. the Windows build has also increased in size? Not critical, just asking.


Regarding Documentation: I am fine with this form, i.e. to split user and developer perspective, and move the former to gh-pages. I now also see the point of setting develop as default then, will do that soon.

Further things I saw here (not related to PR): The Icon does not appear to show up in the About Panel as it is not linked, and the Documentation is of course also only installed/available if the corresponding pdf is available. It would make sense to think about how to improve that at some point.

**Review by yannikschaelte — APPROVED** [review]

File: README.md
Comment (yannikschaelte) [inline README.md]: It's just some info, right? One could add that directly in here.

File: README.md
Comment (yannikschaelte) [inline README.md:3]: add link to https://yannikschaelte.github.io/MiniPlaner/ ?

File: README.md
Comment (yannikschaelte) [inline README.md]: add minimum version? 3.0.4 should be fine.

File: README.md
Comment (mkroening) [reply README.md]: Actually it's more than that.

Those are package scripts for MSYS2. Those are the preferred method of building releases. Currently, they reside in the `impl` branch of that repository, as they depend on this to be merged first.

Package scripts should be separate from the project itself as they are specific to the distribution. Once a new release of MiniPlaner has been issued, I'll publish a package script in the AUR for example.

File: README.md
Comment (mkroening) [reply README.md]: I think it would make more sense to enforce a minimum version via meson and not specify the version in more places, as it would be hard to keep up with those.

What do you think?

**Comment by mkroening** [comment]
I just added hints regarding explicitly setting a wxWidgets config tool and static compilation to the readme.

Regarding your questions:

> * I don't currently have access to Windows, but can for the moment someone confirm that it works on Windows 8 or 10?

I tested this on MSYS2 on Windows 10 which is why this took so long (mainly fiddling around with resource compilation).

> * I have a binary size of 18MB. On Windows, it used to be 7MB. It is probably just the Linux build, or has something in the linking changed s.t. the Windows build has also increased in size? Not critical, just asking.

That's the same size I get on Linux when configuring the build directory with default options. Stripping reduces the file size to a tenth (1,868,328 B).

On Windows when building `mingw-w64-x86_64-miniplaner-git-static` via the package script in MiniPlaner-MINGW the size is 8,879,104 B. Main factors are that the package builds the binary statically (as noted the suffix of the package name) and that makepkg strips the binary before packaging by default. Other options when configuring the build directory should matter as well but not as much.

Where exactly is the icon missing? It is not expected to show on the file like in Windows. For good integration with desktop environments I just opened #35.

I also opened #36.

File: README.md
Comment (mkroening) [reply README.md]: Would you rather have this repository in your GitHub namespace?

Comment (yannikschaelte) [reply README.md]: Ah ok. Then it of course makes sense to keep things in a different repository. If it's about distribution-specificity only, one could also create corresponding branches, I don't have experience on that though.

No need to move it to my namespace I think.

File: README.md
Comment (yannikschaelte) [reply README.md]: agreed

**Comment by yannikschaelte** [comment]
> Where exactly is the icon missing?

In the AboutPanel, it would be again (just for visualization though, not of any practical use). Read from relative path, thus not included.

**Comment by mkroening** [comment]
Ah, I see! I just openend #37 to track that.

File: README.md
Comment (mkroening) [reply README.md]: It makes sense to have those as separate projects for two reasons:
* The build scripts don't share any code with the project code.
* The build scripts are being maintained by whomever is feeling responsible for that platform. Fedora users could maintain RPMs, Debian users could maintain debs, I could maintain the Arch and MSYS2 packages, as they actually use the same package manager.

The project would provide a ready-to-use distribution for Windows (based on my MSYS2 package scripts). We could also look into providing official [flatpaks](https://flatpak.org/), [snaps](https://snapcraft.io/) or [AppImages](https://appimage.org/) in the future. In theory, we could provide distribution specific packages as well, but I'd rather leave that task to users of the corresponding distribution.

Comment (yannikschaelte) [reply README.md]: > It makes sense to have those as separate projects for two reasons:
> 
> * The build scripts don't share any code with the project code.
> * The build scripts are being maintained by whomever is feeling responsible for that platform. Fedora users could maintain RPMs, Debian users could maintain debs, I could maintain the Arch and MSYS2 packages, as they actually use the same package manager.
> 
> The project would provide a ready-to-use distribution for Windows (based on my MSYS2 package scripts). We could also look into providing official [flatpaks](https://flatpak.org/), [snaps](https://snapcraft.io/) or [AppImages](https://appimage.org/) in the future. In theory, we could provide distribution specific packages as well, but I'd rather leave that task to users of the corresponding distribution.

I think probably 95% of users are on Windows, so a standalone Windows distro should suffice for most, other distros would of course be great, however I'm not aware of anyone except us using it. Links to the different distros could then also be put on the gh-pages.

File: README.md
Comment (mkroening) [reply README.md]: Done.

**Comment by mkroening** [comment]
I just added the minimum version requirement to the meson project description.

@yannikschaelte, can you confirm that gh-pages are being built from master instead of develop? I want to be sure I don't break the website when merging this.

**Comment by yannikschaelte** [comment]
built from gh-pages now

**Pull request merged**
